mars/_utils.pyx

```diff
@@ -183,25 +176,50 @@ cdef inline _extract_range_index_attr(object range_index, str attr):
         return getattr(range_index, '_' + attr)
 
 
-cdef h_pandas_index(ob):
+cdef list tokenize_pandas_index(ob):
+    cdef int start
+    cdef int stop
+    cdef int end
     if isinstance(ob, pd.RangeIndex):
         start = _extract_range_index_attr(ob, 'start')
         stop = _extract_range_index_attr(ob, 'stop')
         step = _extract_range_index_attr(ob, 'step')
         # for range index, there is no need to get the values
-        return h_iterative([ob.name, getattr(ob, 'names', None), slice(start, stop, step)])
+        return iterative_tokenize([ob.name, getattr(ob, 'names', None), slice(start, stop, step)])
     else:
-        return h_iterative([ob.name, getattr(ob, 'names', None), ob.values])
+        return iterative_tokenize([ob.name, getattr(ob, 'names', None), ob.values])
 
 
-cdef h_pandas_series(ob):
-    return h_iterative([ob.name, ob.dtype, ob.values, ob.index])
+cdef list tokenize_pandas_series(ob):
+    return iterative_tokenize([ob.name, ob.dtype, ob.values, ob.index])
 
 
-cdef h_pandas_dataframe(ob):
+cdef list tokenize_pandas_dataframe(ob):
     l = [block.values for block in ob._data.blocks]
     l.extend([ob.columns, ob.index])
-    return h_iterative(l)
+    return iterative_tokenize(l)
+
+
+cdef Tokenizer tokenize_handler = Tokenizer()
+
+base_types = (int, long, float, str, unicode, bytes, complex,
+              type(None), type, slice, date, datetime, timedelta)
+for t in base_types:
+    tokenize_handler.register(t, lambda ob: ob)
+
+for t in (np.dtype, np.generic):
+    tokenize_handler.register(t, lambda ob: repr(ob))
+
+for t in (list, tuple):
+    tokenize_handler.register(t, iterative_tokenize)
 
+tokenize_handler.register(np.ndarray, tokenize_numpy)
+tokenize_handler.register(dict, lambda ob: iterative_tokenize(sorted(list(ob.items()), key=str)))
```

Comment:
Since `iterative_tokenize` can handle dict, is it duplicated to define the same logic here?

Reply:
Should be sorted before call `iterative_tokenize `.